design-proposals/klip-38-variable-substitution.md

```diff
@@ -0,0 +1,302 @@
+SET VAR;       # lists all variables
+SET;           # what to do? lists all server, cli, variables, etc?
+```
+
+Also, the use of single-quotes in `SET` server settings is tedious to type. So after removing them, we'll have two different sets:
+* `SET 'server_property' = 'server_value'`
+* `SET VAR variable = `variable_value'
+
+This looks confusing too. So we have to decide for a new syntax for substitution variables.
+
+Other DBs use different syntax to create user variables (DECLARE, DEFINE, \set, SET). For ksqlDB, I'll choose to use `DEFINE` and `UNDEFINE` which
+is what Oracle SQL*plus uses. Any user coming from this DB will immediately understand the command.  
+This is opened to discussion, but will continue with this for now.
+
+#### Creating and printing substitution variables
+
+The command to interact with substitution variables is `DEFINE`.  
+`DEFINE` will allow us to define a new variable, print the variable value and list all current defined variables.
+
+Syntax:
+```
+DEFINE [name [= 'value']];
+
+Where: 
+  name     is the variable name to define
+  value    is the variable value
+```
+Valid variables names start with a letter or underscore (\_) followed by zero or more alphanumeric characters or underscores.
+
+All variable values need to be wrapped using single quotes. All variables are stored as string values in memory, and when
+replaced, the value without the quotes will be used. ksqlDB parsing will be in charge of determining the format at that
+time.
+
+The reason not to remove single-quotes from values is to allow spaces, and to leave the syntax open for using data types in the future if required.  
+Also, having decide the format at run-time (i.e. `define n = 3.34238`) is tricky as we could lose precision in case of decimals. Better stay with Strings for now.
+
+Define a new variable:
+```
+ksql> DEFINE replicas = '3';
+ksql> DEFINE partitions = '5';
+```
+
+Print a defined variable value:
+```
+ksql> DEFINE replicas;
+DEFINE replicas   = '3'
+```
+
+List all defined variables:
+```
+ksql> DEFINE;
+DEFINE replicas   = '3'
+DEFINE partitions = '5'
+```
+
+#### Deleting substitution variables
+
+The `UNDEFINE` command deletes the variable.
+
+Syntax:
+```
+UNDEFINE name;
+```
+
+### 2. Scope
+
+This variable has a session scope.
+
+Variables will live during the session of a ksqlDB connection, then destroyed when the connection is closed. Only the user
+in that session can interact with the defined substitution variables. No other users can interact with other users' variables.
+
+In CLI, the session is opened when starting the CLI with the `ksql` command, and destroyed when typing `exit`.
+
+In Java API client, the session is opened when `Client.create()` is called, and closed when `Client.close()` is called. Variables
+may be defined using a client method instead of a new syntax. This is to ensure users know the scope starts after the `create()` call.
+
+TBD: Define a Java API method for variable substitution.
+
```

Comment:
Do we have plans to support variable substitution in the UI? The UI handles setting properties differently from the CLI: rather than the SET/UNSET syntax for session properties supported by the CLI, query properties instead have to be set using the dropdown in the UI editor. This divergence from the CLI has confused users in the past and I worry that having a different mechanism for defining variables in the UI will lead to similar confusion.

Reply:
Yes, it's gonna be different in the UI than the CLI. I raised this question to @colinhicks too, and seems we'll have this feature separately from the UI. The reason? We'll have to work on opening temporary sessions between the client/server connections, then send variables to the server and stored them temporary there, then do the replacement in the server. That's more involved work.

We might need to do these changes in the future to make better UX for users.